hashicorp/consul
# [CC-7063] Fetch HCP agent bootstrap config in Link reconciler

### Description
This PR updates the link reconciler to get the value of the management token from HCP by fetching the agent bootstrap configuration. This reuses a lot of the existing logic for fetching the agent bootstrap config, specifically the code that first checks if the relevant files already exist locally on disk.

This PR is best viewed commit-by-commit. The main changes are:

- Moves the config loader code into a separate package to avoid cyclical dependencies
- Adds support for loading just the management token rather than also checking for the server configuration, as the server config is not needed by the link resource.
- Passes the data directory as a dependency to the link controller so that it can check if the management token file exists in the data directory or not.
- Calls the load management token (which will in turn call fetch bootstrap config) in the link reconciler
- Handle unauthorized and forbidden errors so that the link status can give a descriptive error and so fetch bootstrap actually does stop on terminal errors
- Validates the resource ID format as part of the link request to return a more descriptive error 


### Testing & Reproduction steps
Read-write scenario:
1. Create a link with HCP credentials to a read-write cluster.
2. Check that the `hcp-config` directory was created in the data-dir and populated as expected.
3. Inspect the logs for the relevant messages.
4. Send another PUT request to the link using the same values to start another reconcile.
5. Inspect the logs to see that the token was loaded locally and not fetched from HCP.

```
$ ls data-dir/
checkpoint-signature	node-id			raft			serf			server_metadata.json
```

```
$ curl \
--request PUT \
--header "Content-Type: application/json" \
--location "http://127.0.0.1:8500/api/hcp/v2/link/global" \
--data '{
  "data": {
    "resource_id": "<redacted>",
    "client_id": "<redacted>",
    "client_secret": "<redacted>"
  }
}'
```
```
$ ls data-dir/
checkpoint-signature	hcp-config		node-id			raft			serf			server_metadata.json

$ ls data-dir/hcp-config/
hcp-management-token	server-config.json	successful-bootstrap
```

```
2024-01-22T10:18:23.347-0600 [DEBUG] agent.server.controller-runtime: fetching configuration from HCP: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMRZEF0Q5F5DXZQHVASZ3JD3\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
2024-01-22T10:18:23.424-0600 [DEBUG] agent.server.controller-runtime: configuration fetched from HCP and saved on local disk: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMRZEF0Q5F5DXZQHVASZ3JD3\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
```

```
2024-01-22T10:18:23.704-0600 [TRACE] agent.server.controller-runtime: loaded HCP configuration from local disk: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMRZEF0Q5F5DXZQHVASZ3JD3\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
```

Read-only scenario:
1. Create a link with HCP credentials to a read-only cluster.
2. Check that the `hcp-config` directory was NOT created.

```
$ curl \
--request PUT \
--header "Content-Type: application/json" \
--location "http://127.0.0.1:8500/api/hcp/v2/link/global" \
--data '{
  "data": {
    "resource_id": "<redacted>",
    "client_id": "<redacted>",
    "client_secret": "<redacted>"
  }
}'
```
```
$ ls data-dir/
checkpoint-signature	node-id			raft			serf			server_metadata.json
```

```
$ curl http://127.0.0.1:8500/api/hcp/v2/link/global 
{
  "data": {
    "accessLevel": "ACCESS_LEVEL_GLOBAL_READ_ONLY",
…
  },
…  },
  "status": {
    "consul.io/hcp/link": {
      "conditions": [
        {
          "message": "Successfully linked to cluster '<redacted>'",
          "reason": "SUCCESS",
          "state": "STATE_TRUE",
          "type": "linked"
        }
      ],
      "observedGeneration": "01HMRZQ5443F2Z312V1GE3WB7Q",
      "updatedAt": "2024-01-22T16:23:07.978814Z"
    }
  },
  "version": "37"
}
```

### Links
- https://hashicorp.atlassian.net/browse/CC-7063

### PR Checklist
